README.md
## pion-radio-example

> Audio (G.711, PCMU) Streaming Example


### Usage
1. start the server
    ```
    $ make rs (run-server)
    ```

2. shoot audio data (sample.wav) to udp port (1234)
    ```
    $ cd audio
    $ make fs (ffmpeg-streaming)
    ```

3. open the web page
    ```
    open http://localhost:8080/static/rtp-radio.html
    press "start session"
    ```

### Reference
- [scottstensland/websockets-streaming-audio](https://github.com/scottstensland/websockets-streaming-audio)
- [WebSocketAudioPlayer](https://archive.codeplex.com/?p=websocketaudioplayer)
- [mohayonao/audio-worker-node](https://github.com/mohayonao/audio-worker-node)




HISTORY.md
### History for developement

- 2020/07/31
    - now, working with audio demo

```
$ sfu-server -h
Usage of sfu-server:
  -dir string
        base directory (default "static")
  -port int
        port of http server (default 8080)
  -rport int
        port of rtp data (default 1234)
```



static/demo.js
/* eslint-env browser */
const log = msg => {
  document.getElementById('logs').innerHTML += msg + '<br>'
}

var isAgent = false;

window.createSession = async isPublisher => {
  let pc = new RTCPeerConnection({
    iceServers: [
      {
        urls: 'stun:stun.l.google.com:19302'
      }
    ]
  })
  pc.oniceconnectionstatechange = e => log(pc.iceConnectionState)
  pc.onicecandidate = event => {
    if (event.candidate === null) {
      document.getElementById('localSessionDescription').value = btoa(JSON.stringify(pc.localDescription))
    }
  }

  try {
    if (isPublisher) {
      const stream = await navigator.mediaDevices.getUserMedia({ video: true, audio: false })
      pc.addStream(document.getElementById('video1').srcObject = stream)
      const offer = await pc.createOffer()
      await pc.setLocalDescription(offer)
      isAgent = isPublisher;
      console.log(isAgent)
    } else {
      pc.addTransceiver('audio', {'direction': 'recvonly'})
      const offer = await pc.createOffer();
      await pc.setLocalDescription(offer)
  
      pc.ontrack = event => {
        var el = document.getElementById('video1')
        el.srcObject = event.streams[0]
        el.autoplay = true
        el.controls = true
      }
    }
  }
  catch(err) {
    log(err);
  }
  
  window.startSession = async () => {
    const sdp = document.getElementById('localSessionDescription').value
    const { data:response } = await axios.post('/sdp', sdp);
    document.getElementById('remoteSessionDescription').value = response;

    let sd = document.getElementById('remoteSessionDescription').value
    if (sd === '') {
      return alert('Session Description must not be empty')
    }

    try {
      pc.setRemoteDescription(new RTCSessionDescription(JSON.parse(atob(sd))))
    } catch (e) {
      alert(e)
    }
  }

  let btns = document.getElementsByClassName('createSessionButton')
  for (let i = 0; i < btns.length; i++) {
    btns[i].style = 'display: none'
  }

  document.getElementById('signalingContainer').style = 'display: block'
}


sfu-server/main.go
package main

import (
	"fmt"
	"log"
	"net"

	"github.com/pion/rtp"
	"github.com/pion/webrtc/v2"

	"github.com/sikang99/pion-radio-example/internal/signal"
)

func init() {
	log.SetFlags(log.Ldate | log.Ltime | log.Lshortfile)
}

func main() {
	sdpInChan, sdpOutChan, rport := signal.HTTPSDPServer()

	m := webrtc.MediaEngine{}
	m.RegisterDefaultCodecs()

	api := webrtc.NewAPI(webrtc.WithMediaEngine(m))

	rtcConfig := webrtc.Configuration{
		ICEServers: []webrtc.ICEServer{
			{
				URLs: []string{"stun:stun.l.google.com:19302"},
			},
		},
	}

	pubConn, err := api.NewPeerConnection(rtcConfig)
	if err != nil {
		log.Fatalln(err)
	}

	laddr, err := net.ResolveUDPAddr("udp", fmt.Sprintf(":%d", rport))
	if err != nil {
		log.Println(err)
		return
	}

	udp, err := net.ListenUDP("udp", laddr)
	if err != nil {
		log.Println(err)
		return
	}
	defer udp.Close()

	inRTPPacket := make([]byte, 4096) // UDP MTU size = 4096
	n, _, err := udp.ReadFromUDP(inRTPPacket)
	if err != nil {
		log.Fatalln(err)
	}

	packet := &rtp.Packet{}
	err = packet.Unmarshal(inRTPPacket[:n])
	if err != nil {
		log.Fatalln(err)
	}
	log.Println("SSRC:", packet.SSRC)

	localTrack, err := pubConn.NewTrack(webrtc.DefaultPayloadTypePCMU, packet.SSRC, "audio", "spider")
	if err != nil {
		log.Fatalln(err)
	}

	_, err = pubConn.AddTrack(localTrack)
	if err != nil {
		log.Fatalln(err)
	}

	go RecvDataToTrack(udp, localTrack)

	log.Println("Now waiting for audio player connections ...")
	for {
		recvOnlyOffer := webrtc.SessionDescription{}
		signal.Decode(<-sdpInChan, &recvOnlyOffer)

		subConn, err := api.NewPeerConnection(rtcConfig)
		if err != nil {
			log.Println(err)
			continue
		}

		_, err = subConn.AddTrack(localTrack)
		if err != nil {
			log.Println(err)
			continue
		}

		err = subConn.SetRemoteDescription(recvOnlyOffer)
		if err != nil {
			log.Println(err)
			continue
		}

		answer, err := subConn.CreateAnswer(nil)
		if err != nil {
			log.Println(err)
			continue
		}

		err = subConn.SetLocalDescription(answer)
		if err != nil {
			log.Println(err)
			continue
		}

		playAnswer := signal.Encode(answer)
		sdpOutChan <- playAnswer
		//log.Println(playAnswer)
	}
}

func RecvDataToTrack(udp *net.UDPConn, toTrack *webrtc.Track) {
	inRTPPacket := make([]byte, 4096)

	for {
		n, _, err := udp.ReadFrom(inRTPPacket)
		if err != nil {
			log.Println(err)
			break
		}

		packet := &rtp.Packet{}
		err = packet.Unmarshal(inRTPPacket[:n])
		if err != nil {
			log.Println(err)
			break
		}

		err = toTrack.WriteRTP(packet)
		if err != nil {
			log.Println(err)
			break
		}
	}
}


internal/signal/http.go
package signal

import (
	"context"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"strconv"

	"github.com/sikang99/pion-radio-example/internal/common"
)

// createSessionKey is for the internal use to avoid duplaicate keys
func createSessionKey(key string) string {
	return "cojam" + key
}

// Middleware process context before handlers
func Middleware(next http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		sessionid := r.Header.Get("x-cojam-session")
		sess, err := common.GetSession(sessionid)
		if err != nil {
			log.Fatalln(err)
		}

		// to control related workers
		ctx, cancel := context.WithCancel(context.Background())
		defer cancel()

		ctx = context.WithValue(ctx, createSessionKey("session"), sess)
		nextRequest := r.WithContext(ctx)

		next(w, nextRequest)
	}
}

//
func faviconHandler(w http.ResponseWriter, r *http.Request) {
	http.ServeFile(w, r, "./static/favicon.ico")
	log.Println(r.URL.String())
}

// PubHandler process publishers
func PubHandler(w http.ResponseWriter, r *http.Request) {
	v := r.Context().Value("session")
	if v == nil {
		http.Error(w, "Not Authorized", http.StatusUnauthorized)
		return
	}
	sess := v.(common.SessionInfo)
	fmt.Fprintf(w, "PubHandler: "+string(sess.UserID))
	log.Println(r.URL.String())
}

// SubHandler process subscribers
func SubHandler(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "SubHandler: ")
	log.Println(r.URL.String())
}

// MonHandler monitor the server
func MonHandler(w http.ResponseWriter, r *http.Request) {
	fmt.Fprintf(w, "MonHandler: ")
	log.Println(r.URL.String())
}

// HTTPSDPServer starts a HTTP Server that consumes SDPs
func HTTPSDPServer() (chan string, chan string, int) {
	port := flag.Int("port", 8080, "port of http server")
	rport := flag.Int("rport", 1234, "port of rtp data")
	dir := flag.String("dir", "static", "base directory of file server")
	//tout := flag.Int("time", 3, "timeout to serve in Second")
	flag.Parse()

	sdpInChan := make(chan string)
	sdpOutChan := make(chan string)

	http.HandleFunc("/favicon.ico", faviconHandler)
	http.HandleFunc("/pub", Middleware(PubHandler))
	http.HandleFunc("/sub", Middleware(SubHandler))
	http.HandleFunc("/mon", Middleware(MonHandler))

	http.HandleFunc("/sdp", func(w http.ResponseWriter, r *http.Request) {
		log.Printf("/sdp connected from %s", r.Host)
		body, _ := ioutil.ReadAll(r.Body)
		// process request of sdp
		sdpInChan <- string(body)
		// send response of sdp
		fmt.Fprintf(w, <-sdpOutChan)
		log.Println("sent base64 SDP to client")
	})

	// http server for static files
	fs := http.FileServer(http.Dir(*dir))
	http.Handle("/static/", http.StripPrefix("/static/", fs))

	go func() {
		err := http.ListenAndServe(":"+strconv.Itoa(*port), nil)
		if err != nil {
			log.Fatalln(err)
		}
	}()

	log.Println("\nWebRTC SFU example server is started")
	log.Printf("started http and file server on :%d and %s", *port, *dir)
	return sdpInChan, sdpOutChan, *rport
}
